### DISCUSSION TITLE
Compiler Planning

### DISCUSSION THREAD

**Discussion by lyxal** [Vyxal/Vyxal#207] [Rewrite] [answered]
List what you think we need to plan for the actual compiler/transpiler.

One thing per answer. No obvious answers like "remove whitespace"

Note that this discussion is _not_ for the actual planning of each point. This is for suggesting what we need to consider; some points may need extended discussion in more appropriate channels.

**Comment by lyxal** [comment]
Generic element template

Elements all have a corresponding python command. E.g. `+` might turn into

```python
rhs, lhs = pop(stack, 2); stack.append(add(lhs, rhs))
```

We should determine the standard template for elements.

**Comment by lyxal** [comment]
python equivalents for each structure

These are the generic versions of each structure. E.g. The template for an if statement might be:

```python
condition = pop(stack)
if condition:
    # truthy code
else:
    # falsey code
```

This would just be a file in the specs folder

**Comment by lyxal** [comment]
what order elements should be implemented

This'll probably be a quick and simple answer, but it's worth considering

**Reply by razetime** [reply]
same order as existing elements list?

**Reply by chunkybanana** [reply]
Digraphs second

**Comment by lyxal** [comment]
who implements which element

We need to have a way to divvy up the element list between developers.

**Reply by razetime** [reply]
have a check list of each command which people have to update whenever they PR an element

**Reply by chunkybanana** [reply]
Or elements

**Comment by lyxal** [comment]
detailed context specification

We need to properly specify how context objects will work - these help remove the need for global variables

**Comment by chunkybanana** [comment]
Keep the functions

The basic functions that run everything behind the scenes  are 99% functional, and the 1% with bugs should be fairly easy to fix, possibly including `vectorise`.

**Reply by lyxal** [reply]
We need to rewrite the functions to move away from global variables, and to use contexts instead.

Also, we can add tests when writing each function if we start from scratch.

**Comment by ysthakur** [comment]
The project structure, including which directory the tests will be in. They currently cause problems because they're outside the vyxal folder. If they're inside that folder, we can use relative imports to get around the ugly path-inserting jank.

**Comment by ysthakur** [comment]
How much detail we want for the doc-strings for each kind of function, and which format we use (it'll have to be one of the more commonly-used ones, of course).

**Comment by pxeger** [comment, marked as answer]
Done, so closing.
